File: C++/src/longest_repeating_character_replacement.cc
#include "longest_repeating_character_replacement.hpp"
#include <map>

int Solution::characterReplacement(string s, int k) {
  int highest_freq = 1;
  int max_len = 1;
  int left = 0;
  int right = 0;
  map<char, int> freq;

  while (size_t(right) < s.length()) {
    freq[s[right]]++;
    highest_freq = max(highest_freq, freq[s[right]]);

    int num_chars_to_replace = right - left + 1 - highest_freq;
    if (num_chars_to_replace > k) {
      freq[s[left]]--;
      left++;
    }

    max_len = right - left + 1;
    right++;
  }

  return max_len;
}

File: src/sliding_window/longest_repeating_character_replacement.rs
struct Solution;

impl Solution {
    pub fn character_replacement(s: String, k: i32) -> i32 {
        let bytes = s.as_bytes();
        let mut counts = [0; 26];
        let mut left = 0usize;
        let mut max_count = 0;
        let mut best = 0;

        for right in 0..bytes.len() {
            let idx = (bytes[right] - b'A') as usize;
            counts[idx] += 1;
            max_count = max_count.max(counts[idx]);

            while (right - left + 1) as i32 - max_count > k {
                let left_idx = (bytes[left] - b'A') as usize;
                counts[left_idx] -= 1;
                left += 1;
            }

            best = best.max(right - left + 1);
        }

        best as i32
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;
    use rstest::rstest;

    #[rstest]
    #[case(String::from("ABAB"), 2, 4)]
    #[case(String::from("AABABBA"), 1, 4)]
    #[case(String::from("AAAA"), 2, 4)]
    fn test_character_replacement(#[case] s: String, #[case] k: i32, #[case] expected: i32) {
        assert_eq!(Solution::character_replacement(s, k), expected);
    }
}
